subsys/net/lib/downloader/src/transports/http.c

```diff
@@ -0,0 +1,571 @@
+	} else {
+		parse_len = buf_len;
+	}
+
+	for (size_t i = 0; i < parse_len; i++) {
+		dl->config.buf[i] = tolower(dl->config.buf[i]);
+	}
+
+	/* Look for the status code just after "http/1.1 " */
+	p = strnstr(dl->config.buf, "http/1.1 ", parse_len);
+	if (p) {
+		q = strnstr(p, "\r\n", parse_len - (p - dl->config.buf));
+		if (q) {
+			/* Received entire line */
+			p += strlen("http/1.1 ");
+			http->header.status_code = strtoul(p, &q, 10);
+		}
+
+	}
+
+	/* The file size is returned via "Content-Length" in case of HTTP,
+	 * and via "Content-Range" in case of HTTPS with range requests.
+	 */
+	do {
+		if (dl->file_size == 0) {
+			if (http->ranged) {
+				p = strnstr(dl->config.buf, "\r\ncontent-range", parse_len);
+				if (!p) {
+					break;
+				}
+				p = strnstr(p, "/", parse_len - (p - dl->config.buf));
+				if (!p) {
+					break;
+				}
+				q = strnstr(p, "\r\n", parse_len - (p - dl->config.buf));
+				if (!q) {
+					/* Missing end of line */
+					break;
+				}
+			} else { /* proto == PROTO_HTTP */
+				p = strnstr(dl->config.buf, "\r\ncontent-length", parse_len);
+				if (!p) {
+					break;
+				}
+				p = strstr(p, ":");
+				if (!p) {
+					break;
+				}
+				q = strnstr(p, "\r\n", parse_len - (p - dl->config.buf));
+				if (!q) {
+					/* Missing end of line */
+					break;
+				}
+				/* Accumulate any eventual progress (starting offset)
+				 * when reading the file size from Content-Length
+				 */
+				dl->file_size = dl->progress;
+			}
+
+			dl->file_size += atoi(p + 1);
+			LOG_DBG("File size = %u", dl->file_size);
+		}
+	} while (0);
+
+	p = strnstr(dl->config.buf, "\r\nconnection: close", parse_len);
+	if (p) {
+		LOG_WRN("Peer closed connection, will re-connect");
+		http->connection_close = true;
+	}
+
+	if (http->header.has_end) {
+		/* We have received the end of the header.
+		 * Verify that we have received everything that we need.
+		 */
+
+		if (!http->header.status_code) {
+			LOG_ERR("Server response malformed: status code not found");
+			return -EBADMSG;
+		}
+
+		expected_status = (http->ranged || dl->progress) ? 206 : 200;
+		if (http->header.status_code != expected_status) {
+			LOG_ERR("Unexpected HTTP response code %ld", http->header.status_code);
+			return -EBADMSG;
+		}
+
+		if (!dl->file_size) {
+			LOG_ERR("File size not set");
+			return -EBADMSG;
+		}
+
+		return parse_len;
+	}
+
+	q = dl->config.buf + buf_len;
+	/* We are still missing part of the header.
+	 * Return the lines (in number of bytes) that we have parsed.
+	 */
+	while (q > dl->config.buf && (*q != '\r') && (*q != '\n')) {
+		q--;
+	}
+
+	/* Keep \r and \n in the buffer in case it is part of the header ending. */
+	while (*(q - 1) == '\r' || *(q - 1) == '\n') {
+		q--;
+	}
+
+	parse_len = (q - dl->config.buf);
+
+	return parse_len;
+}
+
+/* Returns:
+ * Length of data payload left to process on success
+ * Negative errno on error.
+ */
+static int http_parse(struct downloader *dl, size_t len)
```

Comment:
Have you checked if there is already an http parser in Zephyr?

Reply:
There is, I'll leave it as a future task to see if that can be used. 
https://docs.zephyrproject.org/apidoc/latest/parser_8h.html